### PR TITLE
Allow autodec skip; fix audit problem

### PR DESCRIPTION
**Pull request by kylebaron**
# Summary 

- Implements an `autodec` escape hatch, allowing users to list variables to NOT declare in case they are getting picked up
  - This is set in `$ENV MRGSOLVE_AUTODEC_SKIP = c(...)`
  - Some documentation was added to `autodec_clean()` since that was getting more complicated; this is not exported and the docs are for people reading the code only

- Also: I'm just seeing that there is an auditing problem that was introduced with `nm-vars` plugin; basically, we need to skip auditing ODEs for standard models when the `nm-vars` plugin is in play
  - Added `mread.env` slots (logical) indicating whether or not `nm-vars` are in play
  - This is checked prior to standard auditing
  - I plan to harmonize this a bit more down the road ... less piecemeal

- Added a check for reserved names when `nm-vars` in invoked along with `autodec`; this wasn't in place before
  - For example, you should be able to use `A` when `nm-vars` isn't in play; but when it is, it should be reserved and error when `A` is declared
  - This most recent check picks up `A` when it is found by `autodec`

### CODE REVIEW

File: R/modspec.R [hunk 1]
```diff
@@ -631,6 +631,9 @@ parse_env <- function(spec, incoming_names = names(spec),project,ENV=new.env())
   mread.env$annot <- vector("list", n)
   mread.env$ode   <- vector("list", n)
   mread.env$audit_dadt <- FALSE
+  mread.env$`using_nm-vars` <- FALSE
+  # TODO:
+  #mread.env$using_autodec <- FALSE
```
Comment (kylebaron) [inline R/modspec.R:636]: Hold off on setting this for now.

File: R/modspec.R [hunk 2]
```diff
@@ -764,6 +767,13 @@ autodec_vars <- function(code, blocks = NULL) {
...
 #' Clean up `autodec` candidates
 #' 
+#' @param vars candidates
+#' @param rdefs compartments and parameters that will be implemented with 
+#' C++ pre-processor defines
+#' @param build the model `build` object; this contains variables that will 
+#' be globally declared
+#' @param skip additional names to scrub
+#' 
 #' @details
```
Comment (kylebaron) [inline R/modspec.R:777]: internal docs

File: R/mread.R [hunk 3]
```diff
@@ -369,15 +369,22 @@ mread <- function(model, project = getOption("mrgsolve.project", getwd()),
...
+    auto_skip <- cvec_cs(ENV[["MRGSOLVE_AUTODEC_SKIP"]])
     autov <- autodec_vars(spec, blocks = auto_blocks)
-    autov <- autodec_clean(autov, rdefs = rd, build = build)
+    autov <- autodec_clean(
+      autov, 
+      rdefs = rd, 
+      build = build, 
+      skip = auto_skip
+    )
+    autodec_nm_vars(autov, mread.env)
```
Comment (kylebaron) [inline R/mread.R:383]: This is the new check that looks for variables that we're trying to autodec but are reserved when `nm-vars` is invoked.

File: R/mread.R [hunk 4]
```diff
@@ -423,11 +430,14 @@ mread <- function(model, project = getOption("mrgsolve.project", getwd()),
...
   
   # This must come after audit
+  # TODO: harmonize with other audit process
   if(!has_name("ODE", spec)) {
     spec[["ODE"]] <- "DXDTZERO();"
-  } else if(mread.env[["audit_dadt"]]) {
-    spec[["ODE"]] <- c(spec[["ODE"]], ode)
-    audit_spec(x, spec, warn = warn)
+  } else {
+    if(mread.env[["audit_dadt"]] && !mread.env[["using_nm-vars"]]) {
```
Comment (kylebaron) [inline R/mread.R:437]: This fixes the audit issue that we found later on; I have on TODO list to harmonize with other audits.

File: R/nm-mode.R [hunk 5]
```diff
@@ -118,6 +118,22 @@ audit_nm_vars <- function(x, param, init, build, nmv, env) {
   return(invisible(TRUE))
 }
 
+autodec_nm_vars <- function(x, env) {
```
Comment (kylebaron) [inline R/nm-mode.R:121]: This check picks up what was missed when handling `nm-vars`.

File: tests/testthat/test-modspec.R [hunk 6]
```diff
@@ -521,3 +521,19 @@ test_that("autodec models with nm-vars", {
   expect_equal(cpp$var, c("km","err", "cl", "v2", "ka", "CP"))
   expect_equal(cpp$context, c("main", "table",  rep("auto", 4)))
 })
+
+test_that("autodec variables can be skipped", {
```
Comment (kylebaron) [inline tests/testthat/test-modspec.R:525]: Undocumented.

File: R/mread.R [hunk 7]
```diff
@@ -369,15 +369,22 @@ mread <- function(model, project = getOption("mrgsolve.project", getwd()),
       env = mread.env
     )
   }
-  
   # autodec
   if("autodec" %in% names(plugin)) {
     auto_blocks <- c("PREAMBLE", "MAIN", "PRED", "ODE", "TABLE")
+    auto_skip <- cvec_cs(ENV[["MRGSOLVE_AUTODEC_SKIP"]])
```
Comment (kylebaron) [inline R/mread.R:375]: Check if there are any variables to whitelist from `autodec`

**Review by timwaterhouse — APPROVED** [review]
This looks good to me. I don't have any big ideas about how to handle this any other way for now.

**Comment by kylebaron** [comment]
Thanks, @timwaterhouse; let's give this a try in the wild and see what happens. Even if something needs to be adjusted down the road, I'm pretty happy with it ... didn't think it would get this far.

**Pull request merged**